command/kv_rollback.go

```diff
@@ -0,0 +1,258 @@
+  is 5 and the rollback version is 2, the data from version 2 will become
+  version 6.
+
+      $ vault kv rollback -version=2 secret/foo
+
+  Additional flags and more advanced use cases are detailed below.
+
+` + c.Flags().Help()
+	return strings.TrimSpace(helpText)
+}
+
+func (c *KVRollbackCommand) Flags() *FlagSets {
+	set := c.flagSet(FlagSetHTTP | FlagSetOutputFormat)
+
+	// Common Options
+	f := set.NewFlagSet("Common Options")
+
+	f.IntVar(&IntVar{
+		Name:   "version",
+		Target: &c.flagVersion,
+		Usage:  `Specifies the version number that should be made current again.`,
+	})
+
+	return set
+}
+
+func (c *KVRollbackCommand) AutocompleteArgs() complete.Predictor {
+	return nil
+}
+
+func (c *KVRollbackCommand) AutocompleteFlags() complete.Flags {
+	return c.Flags().Completions()
+}
+
+func (c *KVRollbackCommand) Run(args []string) int {
+	f := c.Flags()
+
+	if err := f.Parse(args); err != nil {
+		c.UI.Error(err.Error())
+		return 1
+	}
+
+	var version *int
+	f.Visit(func(fl *flag.Flag) {
+		if fl.Name == "version" {
+			version = &c.flagVersion
+		}
+	})
+
+	args = f.Args()
+
+	switch {
+	case len(args) != 1:
+		c.UI.Error(fmt.Sprintf("Invalid number of arguments (expected 1, got %d)", len(args)))
+		return 1
+	case version == nil:
+		c.UI.Error(fmt.Sprintf("Version flag must be specified"))
+		return 1
+	case c.flagVersion <= 0:
+		c.UI.Error(fmt.Sprintf("Invalid value %d for the version flag", c.flagVersion))
+		return 1
+	}
+
+	var err error
+	path := sanitizePath(args[0])
+
+	client, err := c.Client()
+	if err != nil {
+		c.UI.Error(err.Error())
+		return 2
+	}
+
+	mountPath, v2, err := isKVv2(path, client)
+	if err != nil {
+		c.UI.Error(err.Error())
+		return 2
+	}
+
+	if !v2 {
+		c.UI.Error(fmt.Sprintf("K/V engine mount must be version 2 for rollback support"))
+		return 2
+	}
+
+	path = addPrefixToVKVPath(path, mountPath, "data")
+	if err != nil {
+		c.UI.Error(err.Error())
+		return 2
+	}
+
+	// First, do a read to get the current version for check-and-set
+	var meta map[string]interface{}
+	{
+		secret, err := kvReadRequest(client, path, nil)
+		if err != nil {
+			c.UI.Error(fmt.Sprintf("Error doing pre-read at %s: %s", path, err))
+			return 2
+		}
+
+		// Make sure a value already exists
+		if secret == nil || secret.Data == nil {
+			c.UI.Error(fmt.Sprintf("No value found at %s", path))
+			return 2
+		}
+
+		// Verify metadata found
+		rawMeta, ok := secret.Data["metadata"]
+		if !ok || rawMeta == nil {
+			c.UI.Error(fmt.Sprintf("No metadata found at %s; rollback only works on existing data", path))
+			return 2
+		}
+		meta, ok = rawMeta.(map[string]interface{})
+		if !ok {
+			c.UI.Error(fmt.Sprintf("Metadata found at %s is not the expected type (JSON object)", path))
+			return 2
+		}
+		if meta == nil {
+			c.UI.Error(fmt.Sprintf("No metadata found at %s; rollback only works on existing data", path))
+			return 2
+		}
+
+		// Verify current data found
```

Comment:
I'm not sure this step is necessary. If the key has been marked deleted there wouldn't be any data returned, but we should still allow a rollback on that key.

Reply:
Good point.